lib/shopping_cart.rb

```diff
@@ -0,0 +1,49 @@
+require './lib/product'
+
+class ShoppingCart
+  attr_reader :name, :capacity, :products, :total_number_of_products, :products_by_category
+
+  def initialize(name, capacity)
+    @name = name
+    @capacity = capacity.to_i
+    @products = []
+    @details = {name: @name, capacity: @capacity}
+    @total_number_of_products = 0
+    @products_by_category = {}
+  end
+
+  def add_product(product)
+    @products << product
+    @total_number_of_products += product.quantity
+  end
+
+  def is_full?
+    if @total_number_of_products >= @capacity
+      true
+    else
+      false
+    end
+  end
+
+  def percentage_occupied
+    #i have no idea why this was necessary, but despite the same math it only worked in this format
+    (100.0 / @capacity * @total_number_of_products).round(2)
+  end
+
+  def sorted_products_by_quantity
+    self.products.sort_by {|product| - product.quantity}
+  end
+
+  def product_breakdown#almost done
+    self.products.map do |product|
+    @products_by_category = {
+      product.category => product
+    }
+    end
+  end
```

Comment:
You're super close with this one. Right now, you're only having one category associate with only one product. It looks like you didn't get to the `products_by_category` method, but if you had that, it may come in handy. You could also look into using the `group_by` enum.